health/writer/zcDestination.go

```diff
@@ -0,0 +1,168 @@
+	SystemName string
+	// Metadata that will be added to you datapoints.
+	// There is also some metadata that will be present in datapoint by defualt: source-type
+	Metadata map[string]string
+}
+
+// ZCDestination outputs health data to zenoss cloud endpoint
+type ZCDestination struct {
+	Endpoint *endpoint.Endpoint
+	Config   *ZCDestinationConfig
+}
+
+// Register takes target, builds model and pushes it to preconfigured ZC endpoint
+func (d *ZCDestination) Register(ctx context.Context, target *target.Target) error {
+	model := &zpb.Model{}
+	model.Timestamp = time.Now().UnixNano() / int64(time.Millisecond)
+	model.Dimensions = d.buildTargetDimensions(target.ID, target.Type)
+	model.MetadataFields = d.buildTargetMetadata()
+
+	result, err := d.Endpoint.PutModels(ctx, &zpb.Models{Models: []*zpb.Model{model}})
+	if err != nil {
+		return err
+	}
+
+	log.GetLogger().Debug().Msgf("Models push statistic: succeed: %d, failed: %d", result.Succeeded, result.Failed)
+	return nil
+}
+
+// Push takes health, builds metrics and pushes them to preconfigured ZC endpoint
+func (d *ZCDestination) Push(ctx context.Context, health *target.Health) error {
+	canonicalMetrics := d.buildCanonicalMetrics(health)
+
+	var cmpMetrics []*zpb.CompactMetric
+	if d.Config.UseCompact {
+		compactMetrics, failedMetrics := d.Endpoint.ConvertMetrics(ctx, canonicalMetrics)
+		cmpMetrics = compactMetrics
+		if len(failedMetrics) > 0 {
+			log.GetLogger().Debug().Msgf("Failed to register canonical metrics: %v", failedMetrics)
+			canonicalMetrics = failedMetrics
+		}
+		return nil
+	}
+
+	result, err := d.Endpoint.PutMetrics(ctx, &zpb.Metrics{
+		Metrics:        canonicalMetrics,
+		CompactMetrics: cmpMetrics,
+	})
+	if err != nil {
+		return err
+	}
+	log.GetLogger().Debug().Msgf("Metrics push statistic: succeed: %d, failed: %d", result.Succeeded, result.Failed)
+	return nil
+}
+
+func (d *ZCDestination) buildCanonicalMetrics(health *target.Health) []*zpb.Metric {
+	metrics := make([]*zpb.Metric, 0)
+	for mID, mValue := range health.Metrics {
+		metrics = append(metrics, d.buildMetric(
+			health.TargetID, health.TargetType, mID, mValue,
+		))
+	}
+
+	for cID, cValue := range health.Counters {
+		metrics = append(metrics, d.buildMetric(
+			health.TargetID, health.TargetType, cID, float64(cValue),
+		))
+	}
+
+	return metrics
+}
+
+func (d *ZCDestination) buildMetric(targetID, targetType, metricID string, value float64) *zpb.Metric {
+	metric := &zpb.Metric{}
+	metric.Timestamp = time.Now().UnixNano() / int64(time.Millisecond)
+	metric.Metric = metricID
+	metric.Value = value
+	metric.Dimensions = d.buildTargetDimensions(targetID, targetType)
+	metric.MetadataFields = d.buildTargetMetadata()
+	return metric
+}
+
+func (d *ZCDestination) buildTargetDimensions(targetID, targetType string) map[string]string {
+	dims := make(map[string]string)
+	dims[utils.TargetKey] = targetID
+	dims[utils.TargetTypeKey] = targetType
+	dims[utils.SystemTypeKey] = d.Config.SystemType
```

Comment:
Shouldn't target ID & type suffice to make the metric unique? system type & key would be metadata

Reply:
But you can have multiple things to send the health data under the same tenant.
e.g. I can have two zdm's on different hosts but under the same tenant. They both can have target of type "datasink" and target ID "zenoss" so, we need something else to make a difference. Maybe system-type is not necessary to have in dimensions but at least something like system name so, I can have one more dimension like "zdm-test" and "zdm-prod" - just for example.

but we can manage it through the policy anyway so, I can move it if you wish